File: apps/api/app/routes/cards.py
from __future__ import annotations

import json
from typing import Literal

from fastapi import APIRouter, HTTPException, Request, Response

from app.core.schemas import (
    StudyCardListResponse,
    StudyCardPlacementRequest,
    StudyCardPublic,
    StudyCardSaveRequest,
    StudyCardUpdateRequest,
)

router = APIRouter(prefix="/api/cards", tags=["cards"])


def card_from_row(row) -> StudyCardPublic:
    return StudyCardPublic(
        id=row["id"],
        session_id=row["session_id"],
        card_type=row["card_type"],
        source_action_id=row["source_action_id"],
        source_message_id=row["source_message_id"],
        content=json.loads(row["content_json"]),
        folder_id=row["folder_id"],
        created_at=row["created_at"],
        saved_at=row["saved_at"],
        deferred_at=row["deferred_at"],
    )


@router.get("", response_model=StudyCardListResponse)
def list_cards(
    request: Request,
    card_type: Literal["knowledge_card", "problem_card"] | None = None,
) -> StudyCardListResponse:
    rows = request.app.state.sessions.list_cards()
    if card_type:
        rows = [row for row in rows if row["card_type"] == card_type]
    return StudyCardListResponse(cards=[card_from_row(row) for row in rows])


@router.post("/{card_id}/save", response_model=StudyCardPublic)
def save_card(card_id: str, payload: StudyCardSaveRequest, request: Request) -> StudyCardPublic:
    try:
        row = request.app.state.sessions.save_card(
            card_id,
            session_id=payload.session_id,
            folder_id=payload.folder_id,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="卡片不存在") from exc
    except PermissionError as exc:
        raise HTTPException(status_code=400, detail="卡片不属于当前会话") from exc
    return card_from_row(row)


@router.post("/{card_id}/copy", response_model=StudyCardPublic, status_code=201)
def copy_card(
    card_id: str,
    payload: StudyCardPlacementRequest,
    request: Request,
) -> StudyCardPublic:
    try:
        row = request.app.state.sessions.copy_card(card_id, folder_id=payload.folder_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="卡片或目标文件夹不存在") from exc
    except PermissionError as exc:
        raise HTTPException(status_code=409, detail="待归档卡片不能复制") from exc
    return card_from_row(row)


@router.patch("/{card_id}/move", response_model=StudyCardPublic)
def move_card(
    card_id: str,
    payload: StudyCardPlacementRequest,
    request: Request,
) -> StudyCardPublic:
    try:
        row = request.app.state.sessions.move_card(card_id, folder_id=payload.folder_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="卡片或目标文件夹不存在") from exc
    except PermissionError as exc:
        raise HTTPException(status_code=409, detail="待归档卡片不能移动") from exc
    return card_from_row(row)

@router.put("/{card_id}", response_model=StudyCardPublic)
def update_card(card_id: str, payload: StudyCardUpdateRequest, request: Request) -> StudyCardPublic:
    try:
        row = request.app.state.sessions.update_saved_knowledge_card(
            card_id,
            content=payload.content.model_dump(),
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="卡片不存在") from exc
    except PermissionError as exc:
        raise HTTPException(status_code=409, detail="待归档知识卡片必须先在对话中确认") from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail="只有已归档知识卡片支持修改") from exc
    return card_from_row(row)


@router.delete("", status_code=204)
async def delete_all_cards(request: Request) -> Response:
    if await request.app.state.chat_streams.has_active_streams():
        raise HTTPException(status_code=409, detail="仍有答疑正在生成，请等待完成后再清空全部卡片")
    request.app.state.sessions.delete_all_cards()
    return Response(status_code=204)


@router.delete("/{card_id}", status_code=204)
def delete_card(
    card_id: str,
    request: Request,
) -> Response:
    try:
        request.app.state.sessions.delete_card(card_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="卡片不存在") from exc
    except PermissionError as exc:
        raise HTTPException(status_code=409, detail="待归档卡片不能从全局卡片库删除") from exc
    return Response(status_code=204)

File: apps/web/lib/api/cards.ts
import { API_BASE, JSON_HEADERS, responseError } from "./http";
import type { CardFolder, StudyCard } from "./types";

export async function fetchCards(
  cardType?: "knowledge_card" | "problem_card"
): Promise<StudyCard[]> {
  const params = new URLSearchParams();
  if (cardType) params.set("card_type", cardType);
  const query = params.toString();
  const response = await fetch(`${API_BASE}/api/cards${query ? `?${query}` : ""}`, { cache: "no-store" });
  if (!response.ok) throw await responseError(response, "学习卡片加载失败");
  const payload = await response.json();
  return payload.cards;
}

export async function saveCard(
  cardId: string,
  sessionId: string,
  folderId?: string | null
): Promise<StudyCard> {
  const response = await fetch(`${API_BASE}/api/cards/${cardId}/save`, {
    method: "POST",
    headers: JSON_HEADERS,
    body: JSON.stringify({ session_id: sessionId, folder_id: folderId || null })
  });
  if (!response.ok) throw await responseError(response);
  return response.json();
}

export async function fetchCardFolders(): Promise<CardFolder[]> {
  const response = await fetch(`${API_BASE}/api/card-folders`, { cache: "no-store" });
  if (!response.ok) throw await responseError(response, "卡片文件夹加载失败");
  const payload = await response.json();
  return payload.folders;
}

export async function createCardFolder(input: {
  name: string;
  parent_id?: string | null;
}): Promise<CardFolder> {
  const response = await fetch(`${API_BASE}/api/card-folders`, {
    method: "POST",
    headers: JSON_HEADERS,
    body: JSON.stringify(input)
  });
  if (!response.ok) throw await responseError(response);
  return response.json();
}

export async function updateCardFolder(
  folderId: string,
  input: { name?: string; parent_id?: string | null }
): Promise<CardFolder> {
  const response = await fetch(`${API_BASE}/api/card-folders/${folderId}`, {
    method: "PATCH",
    headers: JSON_HEADERS,
    body: JSON.stringify(input)
  });
  if (!response.ok) throw await responseError(response);
  return response.json();
}

export async function deleteCardFolder(folderId: string) {
  const response = await fetch(`${API_BASE}/api/card-folders/${folderId}`, {
    method: "DELETE"
  });
  if (!response.ok) throw await responseError(response);
}

export async function moveCard(cardId: string, folderId: string): Promise<StudyCard> {
  const response = await fetch(`${API_BASE}/api/cards/${cardId}/move`, {
    method: "PATCH",
    headers: JSON_HEADERS,
    body: JSON.stringify({ folder_id: folderId })
  });
  if (!response.ok) throw await responseError(response);
  return response.json();
}

export async function copyCard(cardId: string, folderId: string): Promise<StudyCard> {
  const response = await fetch(`${API_BASE}/api/cards/${cardId}/copy`, {
    method: "POST",
    headers: JSON_HEADERS,
    body: JSON.stringify({ folder_id: folderId })
  });
  if (!response.ok) throw await responseError(response);
  return response.json();
}

export async function deleteCard(cardId: string) {
  const response = await fetch(`${API_BASE}/api/cards/${cardId}`, {
    method: "DELETE"
  });
  if (!response.ok) throw await responseError(response);
}

export async function deleteAllCards() {
  const response = await fetch(`${API_BASE}/api/cards`, {
    method: "DELETE"
  });
  if (!response.ok) throw await responseError(response);
}
